odoo/http.py

```diff
@@ -583,12 +583,15 @@ def build_controllers():
 
         for top_ctrl in highest_controllers:
```

Comment:
`highest_controllers` are only looked up in the loaded `modules`, so this doesn't look like it actually works for module-less controllers? `Controller.__init_subclass__` will shunt them into `children_classes['']`, from which they won't be retrieved.

So maybe the controllers associated with an empty module (`''`) should always be retrieved here, and any controller subclass with an empty module should be allowed by `is_valid`?

Reply:
Actually I didn't challenge what was `is_valid` used for during the refactor. It feels like we don't want to allow the module inheritance mechanism for non-module-defined controllers. In the latest diff I yield them separately from the other controllers meaning that those controllers are standalone, they cannot be extended anywhere.

Reply:
I mean we basically have to reply to the three following questions:

1) Can non-module-defined controllers extend other non-module-defined ones ?
2) Can module-defined controllers extend non-module-defined ones ?
3) Can non-module-defined controllers extend module-defined ones ?

With the latest diff the answer is No to the three questions, non-module-defined controllers are outside any inheritance mechanism.

Reply:
Sure, I don't care either way.

Current code still seems odd though:

- module-less controllers are not instantiated
- the fallback seems unnecessary, the `highest_controllers` necessarily have one of the selected modules and thus won't be excluded by `get_leaf_classes`

No?

Reply:
Thanks to you last statement "le franc est tombé", indeed the fallback is now unnecessary. +1 for the non-instanciated controllers. I'm fixing both